pkg/diff/diff.go

```diff
@@ -0,0 +1,245 @@
+type changesByPath []archive.Change
+
+func (c changesByPath) Less(i, j int) bool { return c[i].Path < c[j].Path }
+func (c changesByPath) Len() int           { return len(c) }
+func (c changesByPath) Swap(i, j int)      { c[j], c[i] = c[i], c[j] }
+
+// Diff produces a tar archive containing the differences between two filesystems
+func Diff(op trace.Operation, newDir, oldDir string, pathspec map[string]string, data bool) (io.ReadCloser, error) {
+
+	changes, err := archive.ChangesDirs(newDir, oldDir)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Sort(changesByPath(changes))
+
+	spec := processPathSpec(op, pathspec)
+
+	// prune excluded paths from change set
+	for i, change := range changes {
+		if skip(op, change.Path, spec) {
+			changes = append(changes[:i], changes[i+1:]...)
+		}
+	}
+
+	return Tar(op, newDir, changes, spec, data)
+}
+
+func Tar(op trace.Operation, dir string, changes []archive.Change, spec *PathSpec, data bool) (io.ReadCloser, error) {
+	var (
+		err error
+		hdr *tar.Header
+	)
+
+	r, w := io.Pipe()
+	go func() {
+		tw := tar.NewWriter(w)
+
+		for _, change := range changes {
+
+			hdr, err = createHeader(op, dir, change, spec)
+			if err != nil {
+				op.Errorf("Error getting header for change: %s", err.Error())
+				return
+			}
+
+			var f *os.File
+
+			if !data {
+				hdr.Size = 0
+				// include the change type in extended attributes
+				hdr.Xattrs = make(map[string]string)
+				hdr.Xattrs[ChangeTypeKey] = change.Kind.String()
+			}
+
+			_ = tw.WriteHeader(hdr)
+
+			p := path.Join(dir, change.Path)
+			if hdr.Typeflag == tar.TypeReg && hdr.Size != 0 {
+				f, err = os.Open(p)
+				if err != nil {
+					if os.IsPermission(err) {
+						err = nil
+						return
+					}
+					return
+				}
+				if f != nil {
+					_, err = io.Copy(tw, f)
+					_ = f.Close()
+				}
+
+			}
+
+		}
+
+		if err = tw.Close(); err != nil {
+			op.Errorf("Error closing tar writer: %s", err.Error())
+		}
+		if err = w.Close(); err != nil {
+			op.Errorf("Error closing writer: %s", err.Error())
+		}
+		return
+	}()
+	return r, err
+}
+
+// grossly inefficient way to handle exclusion/inclusion of paths
+func skip(op trace.Operation, filePath string, spec *PathSpec) bool {
+	// inefficient way to make sure we don't remove an excluded directory
+	// if it contains a file we explicitly included
+	for k := range spec.Inclusions {
+		if strings.HasPrefix(k, filePath) {
+			return false
+		}
+	}
+
+	p := filePath
+	for p != string(filepath.Separator) {
+		if _, ok := spec.Exclusions[p]; ok {
+			return true
+		}
+		p = path.Dir(p)
+	}
+
+	return false
+}
+
+func createHeader(op trace.Operation, dir string, change archive.Change, spec *PathSpec) (*tar.Header, error) {
+	var hdr *tar.Header
+	timestamp := time.Now()
+
+	switch change.Kind {
+	case archive.ChangeDelete:
+		whiteOutDir := filepath.Dir(change.Path)
+		whiteOutBase := filepath.Base(change.Path)
+		whiteOut := filepath.Join(whiteOutDir, archive.WhiteoutPrefix+whiteOutBase)
```

Comment:
no, `WhiteoutPrefix` is not part of a path, it's part of the filename that is being changed to indicate a delete.